### ISSUE TITLE
Improve how HydPy-L-Land models runoff concentration

### ISSUE THREAD

**Issue by tyralla** [hydpy-dev/hydpy#87]
[L-Land]: https://hydpy-dev.github.io/hydpy/4.0/lland.html
[StateSequence]: https://hydpy-dev.github.io/hydpy/4.0/sequencetools.html#hydpy.core.sequencetools.StateSequence
[FluxSequence]: https://hydpy-dev.github.io/hydpy/4.0/sequencetools.html#hydpy.core.sequencetools.FluxSequence
[QBGZ]: https://hydpy-dev.github.io/hydpy/4.0/lland.html#hydpy.models.lland.lland_states.QBGZ
[QBGA]: https://hydpy-dev.github.io/hydpy/4.0/lland.html#hydpy.models.lland.lland_states.QBGA

[SDG1]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_states.SDG1
[SDG2]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_states.SDG2
[SIG1]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_states.SIG1
[SIG2]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_states.SIG2
[SBG]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_states.SBG

[Return_SG_V1]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_model.Return_SG_V1
[Calc_QBGA_SBG_V1]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_model.Calc_QBGA_SBG_V1
[Update_QDGZ_QBGZ_QBGA_V1]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_model.Update_QDGZ_QBGZ_QBGA_V1
[Update_QDGZ_QBGZ_QBGA_V2]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_model.Update_QDGZ_QBGZ_QBGA_V2

[QDGZ]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_fluxes.QDGZ
[KB]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_derived.KB
[GSBGrad1]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_control.GSBGrad1
[GSBGrad2]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_control.GSBGrad2

Recently, we tried to apply the Ensemble Kalman filter to [L-Land], which did not work well due to the missing state variables for the linear storages of direct runoff, interflow and baseflow.  Like in LARSIM, [L-Land] applies an analytical solution assuming a constant increase or decrease of inflow for the respective time interval.  This approach is, in my opinion, a little confusing and comes with the situation that we need to remember old inflow and outflow as initial states instead of water levels or storage contents.  Thus, [L-Land] defines [QBGZ], [QBGA], and so on as [StateSequence] instead of [FluxSequence] subclasses.

Hence, we decided to switch to an analytical solution assuming a constant inflow that simplifies modifying the states of the different runoff concentration storages externally.

First, we need to introduce new  [StateSequence] subclasses.  My suggestions:
- [SDG1]
- [SDG2]
- [SIG1]
- [SIG2]
- [SBG]

Maybe not the clearest abbreviations, but I think they are relatively consistent with the names of the related fluxes.

Second, [StateSequence] subclasses like [QBGZ] and [QBGA] become [FluxSequence] subclasses.

The ordinary differential equation written in a general manner is $\frac{ds(t)}{dt} = z - \frac{s(t)}{k}$, with $s$ being the storage content in mm, $z$ the stepwise constant inflow in mm/T, and $k$ the concentration time. Its solution is $s(t) = c \cdot e^{-t/k} + k \cdot z$. Applying this solution to the time interval from 0 to 1, we get $s(1) = k \cdot z - (k \cdot z - s(0)) \cdot e^{-1 / k}$. The outflow $a$ for the time interval is $a = s(0) - s(1) + z$.

The following code checks for the correctness of the given solution:

```python
from numpy import exp
from scipy.integrate import ode

k = 10.0
s0 = 5.0
z = 2.0
dt = 1.0

def f_ds(t, s):
    return z - s / k

def f_s(dt):
    return k * z - (k * z - s0) * exp(-dt / k)

integrator = ode(f_ds).set_integrator("zvode", method="bdf", atol=1e-10, rtol=1e-10)
integrator.set_initial_value(s0, 0.0)
numerical = integrator.integrate(dt).real[0]
analytical = f_s(dt)
assert round(numerical, 6) == round(analytical, 6), f"{numerical} vs. {analytical}"
```

In most cases, the implementation of the described solution is straightforward.  To avoid code duplications, I defined the "additional method" [Return_SG_V1], which implements the main logic and is called by the actual "run methods" (e.g. [Calc_QBGA_SBG_V1]) with the suitable parameters.  However, complications arise from interactions with two options implemented in [L-Land] that allow redirecting (potential) inflow of the base flow storage to the direct runoff storage.

For method [Update_QDGZ_QBGZ_QBGA_V1], which redirects the base flow storage's inflow to the direct flow storage when exceeding the groundwater aquifer's limited volume, we can simplify the implementation significantly.  We just need to check for a potential (previously calculated) excess of [SBG] and [QBGA]:

$$\begin{array}{ll}
SBG_{delta} = max \left( SBG_{orig} - GSBMax \cdot VolBMax, 0 \right) \\ 
QBGA_{delta} = max \left( QBGA_{orig} - QBGAMax, 0 \right) 
\end{array}$$

If necessary, we correct [SBG] and [QBGA] and add the total excess to [QDGZ]:

$$\begin{array}{ll}
SBG_{final} = SBG_{final} - SBG_{delta} \\
QBGA_{final} = QBGA_{orig} - QBGA_{delta} \\
QDGZ_{final} = QDGZ_{orig} + SBG_{delta} + QBGA_{delta}
\end{array}$$

One could argue that this approach tends to reduce [QBGA] too little when [SBG] starts to exceed the defined threshold within the current time interval:
<img src="https://user-images.githubusercontent.com/25983587/179901199-2d4dd2d5-d5e0-446f-929f-03a12d0b5957.png" width="300">

However, given the slowness of the base flow and the general accuracy of [L-Land], this should not be of concern.  On the contrary, this approach is more robust than the old one, which resulted in some [artificial oscillations](https://hydpy-dev.github.io/hydpy/4.0/lland_v1_acker_gw_vol.html).  Now, the increase of [QDGZ] follows a [smooth shape](https://hydpy-dev.github.io/hydpy/master/lland_v1_acker_gw_vol.html).

Method [Update_QDGZ_QBGZ_QBGA_V2] redirects (a portion of) the inflow into the storage compartment for base flow into the storage compartments for direct flow due to the groundwater table's too fast rise. The old implementation of this rise-based limitation resulted in even [worse oscillations](https://hydpy-dev.github.io/hydpy/4.0/lland_v1_acker_gw_rise.html) than the old implementation of the volume-based limitation.  Again, we can fix these oscillations, but for the price of relatively complicated implementation.

This is how I understand the basic idea described in the [LARSIM documentation](https://www.larsim.info/dokumentation/LARSIM-Dokumentation.pdf) ($s$ stands for [SBG], $k$ for [KB], $g1$ for [GSBGrad1], and , $g2$ for [GSBGrad2]):

$$s'(t) = \begin{cases}
    1\cdot z - \frac{s(t)}{k}
    &|&
    s'(t) \leq g1
\\
    \frac{g2 - s'(t)}{g2 - g1} \cdot z - \frac{s(t)}{k}
    &|&
    g1 < s'(t) \leq g2
\\
    0 \cdot z- \frac{s(t)}{k}
    &|&
    g2 < s'(t)
\end{cases}$$

Hence, we deal with an interpolation-like approach. As long as $s'(t) \leq g1$, the original linear storage equation applies.  For $s'(t) > g1$, the fraction of the actual and the possible inflow decreases with increasing water level rise.  If $s'(t)$ reaches $g2$ it, automatically becomes zero.  As neither $g2$ nor $k$ can be negative, we can ignore the third case to simplify the equation:

$$s'(t) = \begin{cases}
    z - \frac{s(t)}{k}
    &|&
    s'(t) \leq g1
\\
    \frac{g2 - s'(t)}{g2 - g1} \cdot z - \frac{s(t)}{k}
    &|&
    g1 < s'(t)
\end{cases}$$

I focus on case $g2 > g1$ and leave case $g2 = g1$ for later.

The solution for $s'(t) = z - \frac{s(t)}{k}$ (see above) is monotonous.  Hence, we can safely apply $s(1) = k \cdot z - (k \cdot z - s(0)) \cdot e^{-1 / k}$ whenever $s'(0) \leq 0$.

If $s'(0) > 0$, the storage's outflow ($s(t) / k$) increases during each time interval with a constant inflow ($z$).  So $s'(t)$ decreases over time.  Hence, we can widen the above assertion and safely apply $s(1) = k \cdot z - (k \cdot z - s(0)) \cdot e^{-1 / k}$ whenever $s'(0) \leq g1$.

Unfortunately, things are less straightforward for $s'(0) \leq g1$ because we need to expect that $s'(t)$ becomes smaller than $g1$ sooner or later.

First, we require the analytical solution of  $s'(t) = \frac{g2 - s'(t)}{g2 - g1} \cdot z - \frac{s(t)}{k}$, which is:

$$\begin{array}{ll}
s(t) = c0 \cdot e^{(t \cdot c_1 / k )} - c_2 \\ 
c_1 = \frac{g2 - g1}{g1 - g2 - z} \\
c_2 =  \frac{g2 \cdot k \cdot z}{g1 - g2}
\end{array}$$

Solving for $c$ after substituting $t=0$ yields:

$$c0 = s(0) + c_2$$

Comparing the analytical solution with a numerical approximation requires rearranging the original ODE:

$$s'(t) =\frac{g2 \cdot (s(t) - k \cdot z) - g1 \cdot s}{k \cdot (g1 - g2 - z)}$$

The following script checks the analytical solution similar to the above:

```python
from numpy import exp
from scipy.integrate import ode

k = 10.0
s0 = 5.0
z = 2.0
g1 = 1.0
g2 = 2.0
dt = 1.0

c1 = (g2 - g1) / (g1 - g2 - z)
c2 = (g2 * k * z) / (g1 - g2)

def f_ds(t, s):
    return (g2 * (s - k * z) - g1 * s) / (k * (g1 - g2 - z))

def f_s(dt):
    return (s0 + c2) * exp(dt / k * c1) - c2

integrator = ode(f_ds).set_integrator("zvode", method="bdf", atol=1e-10, rtol=1e-10)
integrator.set_initial_value(s0, 0.0)
numerical = integrator.integrate(dt).real[0]
analytical = f_s(dt)
assert round(numerical, 6) == round(analytical, 6), f"{numerical} vs. {analytical}"
```

So far, we can calculate $s(t)$ for the second case of the ODE.  However, calculating the outflow based on the water balance equation does not work since the actual inflow varies over time.    We can solve this via a quadrature over the analytical ODE solution:

$$\int_{0}^{t} \frac{s(x)}{k} dx = \frac{(s0 + c_2) (e^{c_1 \cdot t / k} - 1)}{c_1} - \frac{c_2 \cdot t}{k}$$

Here is the validation code based on a numerical quadrature:

```python
from numpy import exp
from scipy.integrate import ode, quad

k = 10.0
s0 = 5.0
z = 2.0
g1 = 1.0
g2 = 2.0
dt = 1.0

c1 = (g2 - g1) / (g1 - g2 - z)
c2 = (g2 * k * z) / (g1 - g2)

def f_ds(t, s):
    return (g2 * (s - k * z) - g1 * s) / (k * (g1 - g2 - z))

def f_s(dt):
    return (s0 + c2) * exp(dt / k * c1) - c2

def f_a1(t):
    integrator = ode(f_ds).set_integrator("zvode", method="bdf", atol=1e-10, rtol=1e-10)
    integrator.set_initial_value(s0, 0.0)
    s = integrator.integrate(t).real[0]
    return s / k

def f_a2(t):
    return (s0 + c2) * (exp(c1 * t / k) - 1.0) / c1 - c2 * t / k

numerical = quad(f_a1, 0.0, dt)[0]
analytical = f_a2(dt)
assert round(numerical, 6) == round(analytical, 6), f"{numerical} vs. {analytical}"
```

For determining the time for which $s'(t) = g1$, we can first calculate the corresponding storage content $s^*$:

$$s^* = k \cdot (z - g1)$$

Then we can calculate the searched time point $t^*$:

$$t^* = \frac{k}{c_1} \cdot log \left( \frac{s^* +c_2}{s(0) + c_2} \right)$$

The following code validates the given equation against the results of a Newton-Raphson-based root search:

```python
from numpy import log, exp
from scipy.optimize import newton

k = 10.0
s0 = 5.0
z = 2.0
g1 = 1.0
g2 = 2.0

c1 = (g2 - g1) / (g1 - g2 - z)
c2 = (g2 * k * z) / (g1 - g2)

def f_ds(s):
    return (g2 * (s - k * z) - g1 * s) / (k * (g1 - g2 - z))

def f_s(dt):
    return (s0 + c2) * exp(dt / k * c1) - c2

def f_t():
    s = k * (z - g1)
    return k / c1 * log((s + c2) / (s0 + c2))

numerical = newton(lambda dt: f_ds(f_s(dt)) - g1, 1.0)
analytical = f_t()
assert round(numerical, 6) == round(analytical, 6), f"{numerical} vs. {analytical}"
```

If $t^* \geq 1$, we simply can apply the solution for the second case of the original ODE for $s(1)$.  But for $t^* < 1$, we can apply it for $s(t^* )$ only and then need to switch to the solution for the first case of the ODE for the remaining time ($1 - t^* $).

The equations given above do not work for all allowed (combinations of) parameter values.  One likely scenario is that someone sets [GSBGrad1] and [GSBGrad2] ($g1 = g2$) to the same value, which results in a zero division error.  Then, if $\frac{g2 - s'(t)}{g2 - g1} \cdot z - \frac{s(t)}{k} \geq g1$, the governing ODE simplifies to $s'(t) = g1$ which integrates to:

$$s(t) = s(0) + t \cdot g1$$

We can calculate the corresponding outflow with:

$$\int_{0}^{t} \frac{s(0) + x \cdot g1}{k} dx = t \cdot \frac{g1 \cdot t + 2 \cdot s(0)}{2 \cdot k}$$

The time for which $s'(t) = g1$ is again given by:

$$s^* = k \cdot (z - g1)$$

For the rare case where $g2 = 0$, $s$ remains constant and the inflow into to storage equals its outflow ($s(0) / k$).

Finally, we must consider the special cases where $k = 0$ or $k = \infty $.  If $k = 0$, the storage's outflow is infinitely fast. Hence, the storage must always be zero and method [Update_QDGZ_QBGZ_QBGA_V2] does not need to do anything.  If $k = \infty$, the storage's outflow is always zero.  The ODE ($\frac{g2 - s'(t)}{g2 - g1} \cdot z - \frac{s(t)}{\infty}$) then simplifies to:

$$s'(t) = \frac{g2}{\frac{g2 - g1}{z} + 1}$$

**Comment by tyralla** [comment]
[Calc_QBGA_SBG_V1]: https://hydpy-dev.github.io/hydpy/4.0/lland.html#hydpy.models.lland.lland_model.Calc_QBGA_SBG_V1
[Update_QDGZ_QBGZ_QBGA_V1]: https://hydpy-dev.github.io/hydpy/4.0/lland.html#hydpy.models.lland.lland_model.Update_QDGZ_QBGZ_QBGA_V1
[Update_QDGZ_QBGZ_QBGA_V2]: https://hydpy-dev.github.io/hydpy/4.0/lland.html#hydpy.models.lland.lland_model.Update_QDGZ_QBGZ_QBGA_V2
[Calc_QBGA_SBG_V2]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_model.Calc_QBGA_SBG_V2

After rethinking, I think we should combine [Calc_QBGA_SBG_V1] and [Update_QDGZ_QBGZ_QBGA_V2].  It does not make much sense to apply a "wrong" ODE and then overwrite its result by applying the "right" ODE.  Also, applying two "UPDATE METHODS" in a row can also be confusing.  Hence I merge [Calc_QBGA_SBG_V1] and [Update_QDGZ_QBGZ_QBGA_V2] to the new method [Calc_QBGA_SBG_V2].  So, model developers can still decide to select [Calc_QBGA_SBG_V1] if they want to neglect groundwater rise limitations.

**Comment by tyralla** [comment]
[Calc_QBGA_SBG_V1]: https://hydpy-dev.github.io/hydpy/4.0/lland.html#hydpy.models.lland.lland_model.Calc_QBGA_SBG_V1
[Update_QDGZ_QBGZ_QBGA_V1]: https://hydpy-dev.github.io/hydpy/4.0/lland.html#hydpy.models.lland.lland_model.Update_QDGZ_QBGZ_QBGA_V1
[Update_QDGZ_QBGZ_QBGA_V2]: https://hydpy-dev.github.io/hydpy/4.0/lland.html#hydpy.models.lland.lland_model.Update_QDGZ_QBGZ_QBGA_V2
[Calc_QBGA_SBG_V2]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_model.Calc_QBGA_SBG_V2
[SBG]: https://hydpy-dev.github.io/hydpy/master/lland.html#hydpy.models.lland.lland_states.SBG

After a second rethinking, I think we should also merge the functionality of [Update_QDGZ_QBGZ_QBGA_V1] into [Calc_QBGA_SBG_V2], which allows solving the volume-related limitation without overestimating the outflow and makes everything consistent (and, unfortunately, even a little more complicated).   The complete ODE then becomes:

$$s'(t) = \begin{cases}
    0
    &|&
    s(t) = s_{max}  \land s'(t) > 0
\\
    z - \frac{s(t)}{k}
    &|&
    (s(t) < s_{max} \lor s'(t) \leq 0) \land s'(t) \leq g1
\\
    \frac{g2 - s'(t)}{g2 - g1} \cdot z - \frac{s(t)}{k}
    &|&
    (s_{max} < s(t) \lor s'(t) \leq 0) \land s'(t) > g1
\end{cases}$$

We need to calculate the oint at which the actual storage reaches the highest possible storage ($s_{max}$).  For $s'(0) < g1$, this is:

$$t^* = -k \cdot log \left( \frac{k \cdot z - sm}{k \cdot z - s_{max}} \right) $$

For everything else, it seems we require no further equations (except some variants of the balance equations).

The final version of [Calc_QBGA_SBG_V2] handles the following main-branches:
- no limitation
  - finished
  - volume-limitation
    - finished
- rise-limitation
  - finished
  - volume-limitation
    - finished
  - no limitation
    - finished
    - volume-limitation
      - finished

Additionally, there are much more branches due to the possible special values of $k$, $s_{max}$, $g1$, and $g2$.  I hope, I did not miss a relevant one.

I also implemented a special branch for the case $s0 > s_{max}$, which should never happen, but who knows.  (To ensure $s$ is usually smaller than $s_{max}$, I added a `trim` method to [SBG].

The docstring of [Calc_QBGA_SBG_V2] compares the defined analytical solution with a numerical solution of the original ODE.  It is a little long and complicated, but necessary given the huge number of branches.

**Comment by tyralla** [comment]
Before and after comparison (based on lland_v1)

volume-limitation, before
![grafik](https://user-images.githubusercontent.com/25983587/180928067-96f4287c-9bd9-4f1a-b90e-416571f97cc4.png)

volume-limitation after
![grafik](https://user-images.githubusercontent.com/25983587/180928144-66fbf3f8-6fa2-44d7-ae41-c6d96dc3ede7.png)

rise-imitation before
![grafik](https://user-images.githubusercontent.com/25983587/180928424-4ba59291-f00d-4e9d-bac9-bd210c822774.png)

rise-limitation after
![grafik](https://user-images.githubusercontent.com/25983587/180928300-59df6f4a-2956-47ad-9dd6-19a92bb03117.png)

**Comment by tyralla** [comment]
Implemented by commit 8d55d2b4fb8e94dda64712f630ae0f8f26bcb656.  The results seem satisfactory (see the results above), and I am confident that the massive amount of test cases captures at least the majority of realistic use cases.  However, I still prefer to postpone this change to the next release (5.1).  So we have some time to evaluate it in real applications.
